db/user_js_runner.go

```diff
@@ -0,0 +1,384 @@
+		return nil, base.HTTPErrorf(http.StatusInternalServerError, "Error compiling %s %q: %v", kind, name, err)
+	}
+
+	// Function that runs before every call:
+	runner.Before = func() {
+		if runner.currentDB == nil {
+			panic("javaScriptRunner can't run without a currentCtx or currentDB")
+		}
+	}
+	// Function that runs after every call:
+	runner.After = func(jsResult otto.Value, err error) (interface{}, error) {
+		defer func() {
+			runner.currentDB = nil
+		}()
+		if err != nil {
+			base.ErrorfCtx(context.Background(), base.KeyJavascript.String()+": %s %s failed: %#v", runner.kind, runner.name, err)
+			return nil, runner.convertError(err)
+		}
+		return jsResult.Export()
+	}
+
+	return runner, nil
+}
+
+// Calls a javaScriptRunner's JavaScript function.
+func (runner *userJSRunner) CallWithDB(db *Database, mutationAllowed bool, args ...interface{}) (result interface{}, err error) {
+	runner.currentDB = db
+	runner.mutationAllowed = mutationAllowed
+	ctx := db.Ctx
+	var timeout time.Duration
+	if ctx != nil {
+		if deadline, exists := ctx.Deadline(); exists {
+			timeout = time.Until(deadline)
+			if timeout <= 0 {
+				return nil, sgbucket.ErrJSTimeout
+			}
+		}
+	}
+	runner.SetTimeout(timeout)
+	return runner.Call(args...)
+}
+
+// JavaScript error returned by a userJavaScriptRunner
+type jsError struct {
+	err        error
+	runnerKind string
+	runnerName string
+}
+
+func (jserr *jsError) Error() string {
+	return fmt.Sprintf("%v (while calling %s %q)", jserr.err, jserr.runnerKind, jserr.runnerName)
+}
+
+func (jserr *jsError) Unwrap() error {
+	return jserr.err
+}
+
+var HttpErrRE = regexp.MustCompile(`^HTTP:\s*(\d+)\s+(.*)`)
+
+func (runner *userJSRunner) convertError(err error) error {
+	if err == sgbucket.ErrJSTimeout {
+		return base.HTTPErrorf(408, "Timeout in JavaScript")
+	}
+	// Unfortunately there is no API on otto.Error to get the name & message separately.
+	// Instead, look for the name as a prefix. (See the `ottoResult` function below)
+	str := err.Error()
+	if strings.HasPrefix(str, "HTTP:") {
+		m := HttpErrRE.FindStringSubmatch(str)
+		status, _ := strconv.ParseInt(m[1], 10, 0)
+		message := m[2]
+		if status == http.StatusUnauthorized && (runner.currentDB.user == nil || runner.currentDB.user.Name() != "") {
+			status = http.StatusForbidden
+		}
+		return base.HTTPErrorf(int(status), "%s (while calling %s %q)", message, runner.kind, runner.name)
+	}
+	return &jsError{err, runner.kind, runner.name}
+}
+
+//////// DATABASE CALLBACK FUNCTION IMPLEMENTATIONS:
+
+// Implementation of JS `app.func(name, params)` function
+func (runner *userJSRunner) do_func(funcName string, params map[string]interface{}) (interface{}, error) {
+	return runner.currentDB.CallUserFunction(funcName, params, runner.mutationAllowed)
+}
+
+// Implementation of JS `app.get(docID, docType)` function
+func (runner *userJSRunner) do_get(docID string, docType *string) (interface{}, error) {
+	if err := runner.currentDB.CheckTimeout(); err != nil {
```

Comment:
For my own understanding - when app.get() returns a timeout error, the calling (customer-defined) javascript function has the option to catch that error and not necessarily return the timeout to the client, is that correct?

Reply:
I hadn't considered that. Hmm. 

In principle, it doesn't matter because any native function it calls after that will keep throwing a timeout error, and even Otto's Interrupt handler will panic the next time it's called (and that panic does _not_ get intercepted by JS; it just flies straight out of Otto into our code.)

However, the Context timeout mechanism works by reading a value from a channel. Does that mean that you only get one chance to find out about the timeout, and if you keep going you won't receive another value so the timeout checks won't trigger anymore?

Reply:
WithTimeout is going to close the context's Done channel, so that's safe for multiple reads.